### PR TITLE
[FE] Category page

### PR DESCRIPTION
**Pull request by mwkim0919**

### CODE REVIEW

**Comment by johnp418** [comment]
Let's add `yarn build` command in our travis yml file. Frontend doesn't compile right now

File: frontend/src/utils/categoryUtil.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    };
+    acc[others] = {
+      nameSets: [others],
+      limit: 0,
+    };
+    return acc;
+  }, {});
+};
+
+export const groupAmountByCategory = (transactions: Transaction[]): object => {
```
Comment (johnp418) [inline frontend/src/utils/categoryUtil.ts]: Let's refactor this when we meet.

File: frontend/src/components/Category/CategoryList.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      groupAmountByCategory(selectedTransactions)
+    );
+    console.log(categoryAmountMap);
+    return (
+      <div>
+        <label htmlFor="exampleFormControlSelect1">Date</label>
+        <select
+          className="form-control"
+          id="date"
+          onChange={e => this.selectDate(e)}
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/components/Category/CategoryList.tsx]: Changed.

File: frontend/src/utils/categoryUtil.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  return transactions.reduce((acc: object, transaction: Transaction) => {
+    const key = transaction.categoryName || 'Others';
+    if (acc[key]) {
+      acc[key] += Number(transaction.withdraw);
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/utils/categoryUtil.ts]: Changed.

File: frontend/src/components/Category/CategoryList.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        >
+          {Array.from(months).map((date: string) => {
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/components/Category/CategoryList.tsx]: Changed.

File: frontend/src/utils/categoryUtil.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    resultArray.push(category.parentCategory.name);
+    storeAllSubCategoryNames(category.parentCategory, resultArray);
+  }
+  return resultArray;
+};
+
+export const generateCategoryMaps = (categories: Category[]): object => {
+  return categories.reduce((acc: object, category: Category) => {
+    acc[category.name] = {
+      nameSets: storeAllSubCategoryNames(category, [category.name]).reverse(),
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/utils/categoryUtil.ts]: Changed

File: frontend/src/types/budget.ts [hunk 6]
```diff
@@ -45,8 +45,8 @@ export interface TransactionFormItem {
 export interface Category {
   id: number;
   name: string;
-  categoryLimit: number;
-  parentCategory?: Category;
+  categoryLimit: number | string | null;
+  parentCategory?: Category | null;
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/types/budget.ts]: Changed.

File: frontend/src/components/Category/CategoryItem.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    }
+  }
+  render() {
+    const { displayName, limit, amount } = this.props;
+    const percent = ((amount / (limit || 1)) * 100);
+    const percentString = percent > 100 ? '100%' : percent.toFixed(0) + '%';
+    const limitString = limit ? 'out of $' + limit : '';
+    return (
+      <div className="media text-muted pt-3">
+        <p className="media-body pb-3 mb-0 small lh-125 border-bottom border-gray">
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/components/Category/CategoryItem.tsx]: Removed.

File: frontend/src/utils/categoryUtil.ts (hunk 1 above)
Comment (mwkim0919) [reply frontend/src/utils/categoryUtil.ts]: will discuss later

File: frontend/src/components/Category/CategoryList.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  date: string;
+}
+
+export default class CategoryList extends React.Component<Props, State> {
+  state = {
+    date: '',
+  };
+
+  // @ts-ignore
+  selectDate(e) {
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/components/Category/CategoryList.tsx]: added.

File: frontend/src/utils/categoryUtil.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      limit: category.categoryLimit,
+    };
+    acc['Others'] = {
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/utils/categoryUtil.ts]: changed.

**Pull request merged**